modules/kafka/src/main/java/org/testcontainers/containers/KafkaContainer.java

```diff
@@ -54,40 +56,67 @@ public KafkaContainer withExternalZookeeper(String connectString) {
     }
 
     public String getBootstrapServers() {
-        return String.format("PLAINTEXT://%s:%s", proxy.getContainerIpAddress(), proxy.getFirstMappedPort());
+        if (port == -1) {
+            throw new IllegalStateException("You should start Kafka container first");
+        }
+        return String.format("PLAINTEXT://%s:%s", getContainerIpAddress(), port);
     }
 
     @Override
-    protected void doStart() {
-        String networkAlias = getNetworkAliases().get(0);
-        proxy = new SocatContainer()
-                .withNetwork(getNetwork())
-                .withTarget(KAFKA_PORT, networkAlias)
-                .withTarget(ZOOKEEPER_PORT, networkAlias);
+    @NonNull
+    public synchronized Network getNetwork() {
+        if (super.getNetwork() == null) {
+            // Backward compatibility
+            withNetwork(Network.newNetwork());
+        }
+        return super.getNetwork();
+    }
 
-        proxy.start();
-        withEnv("KAFKA_ADVERTISED_LISTENERS", "BROKER://" + networkAlias + ":9092" + "," + getBootstrapServers());
+    @Override
+    protected void doStart() {
+        withCommand("sleep infinity");
 
-        if (externalZookeeperConnect != null) {
-            withEnv("KAFKA_ZOOKEEPER_CONNECT", externalZookeeperConnect);
-        } else {
+        if (externalZookeeperConnect == null) {
             addExposedPort(ZOOKEEPER_PORT);
-            withEnv("KAFKA_ZOOKEEPER_CONNECT", "localhost:2181");
-            withCommand(
-                "sh",
-                "-c",
-                // Use command to create the file to avoid file mounting (useful when you run your tests against a remote Docker daemon)
-                "printf 'clientPort=2181\ndataDir=/var/lib/zookeeper/data\ndataLogDir=/var/lib/zookeeper/log' > /zookeeper.properties" +
-                    " && zookeeper-server-start /zookeeper.properties" +
-                    " & /etc/confluent/docker/run"
-            );
         }
 
         super.doStart();
     }
 
     @Override
-    public void stop() {
-        Stream.<Runnable>of(super::stop, proxy::stop).parallel().forEach(Runnable::run);
+    @SneakyThrows
+    protected void containerIsStarting(InspectContainerResponse containerInfo) {
+        super.containerIsStarting(containerInfo);
+
+        port = getMappedPort(KAFKA_PORT);
+
+        final String zookeeperConnect;
+        if (externalZookeeperConnect != null) {
+            zookeeperConnect = externalZookeeperConnect;
+        } else {
+            zookeeperConnect = "localhost:" + ZOOKEEPER_PORT;
+
+            // Start ZooKeeper
```

Comment:
WDYT about moving the code to a private method called `startInternalZooKeeper()` and omit the comment? :man_shrugging: